File: tests/fixtures/python/supers.py
import json
from typing import Generic, TypeVar

from repos import AuditLog

T = TypeVar("T")


class Archive:
    label: str = "archive"

    def __init__(self, audit: AuditLog) -> None:
        self.audit = audit

    def store(self, item: int) -> None:
        pass

    def flush(self) -> None:
        pass


class ColdArchive(Archive):
    def __init__(self, audit: AuditLog) -> None:
        super().__init__(audit)

    def store(self, item: int) -> None:
        super().store(item)


class GlacierArchive(ColdArchive, Generic[T]):
    def store(self, item: int) -> None:
        super().store(item + 1)
        super().flush()
        print(super().label)

    def flush(self) -> None:
        def later() -> None:
            super().flush(), "no arguments to find in here"

        later()


class Stamped:
    def store(self, item: int) -> None:
        pass

    def stamp(self) -> None:
        pass


class Mixed(Stamped, ColdArchive):
    def store(self, item: int) -> None:
        super().store(item + 2)
        super().stamp()
        super().flush(), "one base leads to it"


class Left(Archive):
    pass


class Right(Archive):
    def flush(self) -> None:
        pass


class Diamond(Left, Right):
    def flush(self) -> None:
        super().flush(), "Right.flush, which a walk by depth passes"


class Wire(json.JSONEncoder, Archive):
    def store(self, item: int) -> None:
        super().store(item + 3)

    def default(self, o: object) -> object:
        return super().default(o)


class Ledger(Archive):
    def close(self) -> None:
        super().audit.store(0)

File: tests/fixtures/typescript/supers.ts
import { AuditLog } from "./repos";

export class Archive {
  label = "archive";

  constructor(protected audit: AuditLog) {}

  store(item: number): void {}

  flush(): void {}

  toString(): string {
    return this.label;
  }
}

export class ColdArchive extends Archive {
  store(item: number): void {
    super.store(item);
  }
}

export class GlacierArchive extends ColdArchive {
  store(item: number): void {
    super.store(item + 1);
    super.flush();
    [2].forEach((n) => super.store(n));
  }

  flush(): void {
    const plain = {
      toString(): string {
        return "plain " + super.toString();
      },
    };
    void plain;
  }
}

export class Vault {
  open(): Archive {
    return new Archive(new AuditLog());
  }
}

export class ColdVault extends Vault {
  open(): ColdArchive {
    const opened = super.open();
    opened.store(1);
    super.open().store(2);
    return new ColdArchive(new AuditLog());
  }
}
